dali/pipeline/operators/geometric/flip.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/geometric/flip.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+DALI_SCHEMA(Flip)
+    .DocStr(R"code(Converts between various image color models)code")
+        .NumInput(1)
+        .NumOutput(1)
+        .AllowMultipleInputSets()
+        .AddOptionalArg("horizontal",
+                R"code(The color space of the input image)code",
+                1)
+        .AddOptionalArg("vertical",
+                R"code(The color space of the output image)code",
+                0);
+
+template <>
+Flip<CPUBackend>::Flip(const OpSpec& spec)
+    : Operator<CPUBackend>(spec)
+    , _horizontal(spec.GetArgument<int32>("horizontal"))
+    , _vertical(spec.GetArgument<int32>("vertical")) {}
+
+
+enum class FlipAxis {
+  HORIZONTAL,
+  VERTICAL,
+  BOTH
+};
+
+template <FlipAxis D>
+struct FlipImpl {};
+
+template <>
+struct FlipImpl<FlipAxis::HORIZONTAL> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      for (size_t c = 0; c < width; ++c) {
+        std::copy(
+            &input[(r * width + c) * channels],
+            &input[(r * width + c + 1) * channels],
+            &output[(r * width + width - c - 1) * channels]);
+      }
+    }
+  }
+};
+
+template <>
+struct FlipImpl<FlipAxis::VERTICAL> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      std::copy(
+          &input[r * width * channels],
+          &input[(r + 1) * width * channels],
+          &output[(height - r - 1) * width * channels]);
+    }
+  }
+};
+
+template <>
+struct FlipImpl<FlipAxis::BOTH> {
+  template <typename T>
+  static void run(T *output, const T *input, size_t height, size_t width, size_t channels) {
+    for (size_t r = 0; r < height; ++r) {
+      for (size_t c = 0; c < width; ++c) {
+        std::copy(
```

Comment:
This makes an optimistic assumption that compiler does strength reduction for you.